Clean/Gadgets/BLAKE3/ApplyRounds.lean

```diff
@@ -0,0 +1,1068 @@
+- Second round, permute message
+Returns the final state and permuted message.
+-/
+def applyTwoRounds (state : Vector Nat 16) (message : Vector Nat 16) : Vector Nat 16 × Vector Nat 16 :=
+  let state1 := round state message
+  let msg1 := permute message
+  let state2 := round state1 msg1
+  let msg2 := permute msg1
+  (state2, msg2)
+
+/--
+Specification for two rounds that matches the pattern of the full ApplyRounds.Spec.
+-/
+def TwoRoundsSpec (input : Round.Inputs (F p)) (output : Round.Inputs (F p)) : Prop :=
+  let (final_state, final_message) := applyTwoRounds input.state.value (input.message.map U32.value)
+  output.state.value = final_state ∧
+  output.message.map U32.value = final_message ∧
+  output.state.Normalized ∧
+  (∀ i : Fin 16, output.message[i].Normalized)
+
+
+/--
+Two rounds with permute, but with a spec matching the applyRounds pattern.
+-/
+def twoRoundsApplyStyle : FormalCircuit (F p) Round.Inputs Round.Inputs :=
+  twoRoundsWithPermute.weakenSpec TwoRoundsSpec (by
+    -- Prove that twoRoundsWithPermute's spec implies our TwoRoundsSpec
+    intro input output h_assumptions h_spec
+    -- twoRoundsWithPermute.Spec says ∃ mid, roundWithPermute.Spec input mid ∧ roundWithPermute.Spec mid output
+    obtain ⟨mid, h_spec1, h_spec2⟩ := h_spec
+    -- Unpack what each roundWithPermute spec gives us
+    simp only [roundWithPermute] at h_spec1 h_spec2
+    simp only [TwoRoundsSpec, applyTwoRounds]
+
+    -- From h_spec1: mid.state.value = round input.state.value (BLAKE3State.value input.message)
+    -- From h_spec1: BLAKE3State.value mid.message = permute (BLAKE3State.value input.message)
+    -- From h_spec2: output.state.value = round mid.state.value (BLAKE3State.value mid.message)
+    -- From h_spec2: BLAKE3State.value output.message = permute (BLAKE3State.value mid.message)
+
+    constructor
+    · -- Prove: output.state.value = round (round input.state.value (input.message.map U32.value)) (permute (input.message.map U32.value))
+      rw [h_spec2.1, h_spec1.1]
+      -- Need to show BLAKE3State.value mid.message = permute (Vector.map U32.value input.message)
+      -- We have h_spec1.2.2.1: BLAKE3State.value mid.message = permute (BLAKE3State.value input.message)
+      -- Need to show BLAKE3State.value input.message = Vector.map U32.value input.message
+      rw [h_spec1.2.2.1]
+      -- Show BLAKE3State.value input.message = Vector.map U32.value input.message
+      rfl
+    constructor
+    · -- Prove: output.message.map U32.value = permute (permute (input.message.map U32.value))
+      -- We have h_spec2.2.2.1: BLAKE3State.value output.message = permute (BLAKE3State.value mid.message)
+      -- We have h_spec1.2.2.1: BLAKE3State.value mid.message = permute (BLAKE3State.value input.message)
+      -- Use our lemma to convert between the notations
+      rw [← blake3_value_eq_map_value output.message]
+      rw [h_spec2.2.2.1, h_spec1.2.2.1]
+      rw [blake3_value_eq_map_value input.message]
```

Comment:
can we initially `simp_all [blake3_value_eq_map_value]` and forget about the adjustments?

Reply:
These disappeared.